Backport "assumed valid blocks" feature from Bitcoin 0.13

codablock opened the pull request:
This PR backports the "assumed valid blocks" feature from Bitcoin 0.13+ as already discussed in https://www.dash.org/forum/threads/backporting-bitcoin-0-13-performance-improvements.15612/

The PR includes multiple backported commits and 1 custom commit.

**EDIT: As requested, I added original commit IDs to the description. Also, I added a few more commits as I realized that I should not have selected commits on naive commit level but on PR level. I reordered the commits based on PRs and marked the new commits with NEW**

From original PR https://github.com/bitcoin/bitcoin/pull/9053
1. "IBD check uses minimumchain work instead of checkpoints" (original commit: fd46136dfaf68a7046cf7b8693824d73ac6b1caf)
This one introduced the nMinimumChainWork chainparam which is needed by the assumed valid blocks feature.
I updated the nMinimumChainWork param for both the Dash mainnet and the testnet.
2. **NEW** "Remove GetTotalBlocksEstimate and checkpoint tests that test nothing." (original commit: 2082b5574cec783f4ff99941492d92e05680b293)
3. **NEW** "IsInitialBlockDownload no longer uses header-only timestamps." (original commit: e141beb6a9816b7e1e680fb0a8bae16d42a3e557)

From original PR https://github.com/bitcoin/bitcoin/pull/10345
1. "Delay parallel block download until chain has sufficient work" (original commit: e2652002b6011f793185d473f87f1730c625593b)
As written in the forum post, I noticed that block headers and blocks are downloaded at the same time. When the node tries to figure out if an incoming block is an ancestor of the assumed valid block, it needs all the headers up until the assumed valid block. This means, it will always trigger full validation until the required headers are downloaded. As headers and blocks are downloaded at the same time, the headers download is slowed down quite a bit, so it's better to use full bandwidth for headers before we start the actual block download. This commits delays the block download.
2. **NEW** Add timeout for headers sync (original commit: 76f74811c44b5119d5c19364b0594d423248ac0e)

Custom commits:
1. "Add consensusParams to FindNextBlocksToDownload"
I had to add the consensusParams argument as it is not there yet due to other missing backports. This may result in minor conflicts later when other stuff is backported.

From original PR https://github.com/bitcoin/bitcoin/pull/9484
1. "Introduce assumevalid setting to skip presumed valid scripts." (original commit: e440ac7ef3b6f3ad1cd8fc7027cece40413202d9)
This is the actual feature being backported. It introduces the "-assumevalid" cli argument and the defaultAssumeValid chainparam. I've updated defaultAssumeValid for both the Dash mainnet and testnet.

**EDIT: The above commits miss a new testcase being added in the original PR. I'm gonna handle this while fixing the other failing tests.** 

I would like to ask a few reviewers to verify the values used for nMinimumChainWork and defaultAssumeValid, to make sure I chose correct values. This can be done with your own full node and the debug console. I used block 721000 on mainnet and 251000 on testnet. To check if I used correct values, you could for example use the command "getblockhash 721000" and then "getblock XXX" with the result of getblockhash. getblock will also tell you the chainwork of that block.

Currently, it's hard to see the progress of the synchronization with this PR applied, as headers synching is not shown yet. It looks like nothing is happening for some time, until the actual block download starts. To see progress, it's needed to enter the debug console and use "getblockchaininfo", which will tell you how many headers are downloaded.

The current Bitcoin code already fixes this with some enhancements to the Qt wallet. The next step would be to backport some of these changes to enhance the user experience.

codablock commented:
I see at least one test case is failing, gonna dig through it tomorrow.

UdjinM6 commented:
ACK min work/block hashes for e10e3bd/8ec2c93
Would be nice to have references to original commits (hashes) in description.

codablock commented:
I've updated the PR description with original commit IDs. I also added 3 new commits which I missed by looking at commit level instead of PR level.

I've also just noticed a problem with the initial headers sync which was probably unnoticed until now (due to headers and blocks being synced at the same time). I see something around 10 times as much data downloaded as the expected total headers size. This is most likely due to https://github.com/bitcoin/bitcoin/issues/6755, which doesn't have a working fix in Bitcoin right now. I'll try to dig deeper to figure out if I can somehow make a fix for this.

UdjinM6 commented:
Thanks!

re-ack for block hashes in 8242661

>NEW [P2P] Timeout for headers sync (original commit: 76f7481)

I don't see backport for this one in commit list

codablock commented:
Sorry about the missing commit. Also, I copy&pasted the wrong commit title into the description, it's actually "Add timeout for headers sync". I added that commit and pushed.

UdjinM6 commented on src/net_processing.cpp:
```diff
@@ -2605,6 +2609,39 @@ bool SendMessages(CNode* pto, CConnman& connman, std::atomic<bool>& interruptMsg
+        // Check for headers sync timeouts
+        if (state.fSyncStarted && state.nHeadersSyncTimeout < std::numeric_limits<int64_t>::max()) {
+            // Detect whether this is a stalling initial-headers-sync peer
+            if (pindexBestHeader->GetBlockTime() <= GetAdjustedTime() - 24*60*60) {
```
I believe this timeout logic is connected to `nMaxTipAge` value and should be adjusted just like we did on line 2399 earlier

codablock replied:
I'll add an additional commit to fix this to 6*60*60. Can you explain why these values need to be different in Dash? As I understand it, the change from 24 to 6 would make the approximate amount of blocks (144) to line up with that of Bitcoin again. Does 144 blocks have any other special meaning than being a day for Bitcoin?

UdjinM6 replied:
I don't know of any other special meaning for 144 blocks besides the one you mentioned tbh. Same applies for 2 weeks (i.e. 2016 blocks, diff adjustment period). But since we already have this logic (to follow similar number of blocks and to keep some logic for timespan even though it doesn't apply directly due to dynamic diff adjustment) in other parts of Dash legacy code, we simply have to follow it in new parts too to be consistent. Or change this all at once if we decide to do this one day, but this kind of changes should definitely be in a separate PR then.

codablock replied:
Added a new commit now.

UdjinM6 commented on src/validation.cpp:
```diff
@@ -1967,11 +1972,28 @@ bool ConnectBlock(const CBlock& block, CValidationState& state, CBlockIndex* pin
+                //  artificially set the default assumed verified block further back.
+                // The test against nMinimumChainWork prevents the skipping when denied access to any chain at
+                //  least as good as the expected chain.
+                fScriptChecks = (GetBlockProofEquivalentTime(*pindexBestHeader, *pindex, *pindexBestHeader, chainparams.GetConsensus()) <= 60 * 60 * 24 * 7 * 2);
```
... and logic for this one I guess is connected to `nPowTargetTimespan` which is 1 day in Dash

codablock replied:
As I understand the comment above the code, it also protects against bad releases in case someone took a too recent hash as default value just before doing the release. In that case testing would be worth less as only a few blocks would be verified by all testers before release.

UdjinM6 replied:
Hmm, you are probably right about this one and it has nothing to do with `nPowTargetTimespan` as I originally guessed. Let's keep it 2 weeks then.

codablock commented:
Checkpoints seem to be in the process of being removed in Bitcoin. I didn't check the progress of that however, just seen a hint in one of the commit descriptions.
